<a href="https://colab.research.google.com/github/wslock/-D012513A_Python_20230124/blob/main/Practical%20Python%20Specialized%20Bioinformatics%2020230124%20solution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Practical Python Specialized Bioinformatics

#### 24/01/2023

For this exam you are asked to complete the Python code between the 

```
#start code here

#end code here
```

comments. 

A direct question needs to be answered in the text field just below the question.

Don't forget to periodically save your work to your GitHub repository. 

Please write your name and studentnumber in the text box below:

### Name: Warre Slock

### Number: 01903407

### Total points:    / 25

In [1]:
import warnings;
warnings.filterwarnings('ignore');

import pandas as pd

**question 1 (2pts)**

Create a Pandas DataFrame called `patients` that looks like this:

|           | gene_1 | gene_2 |
|-----------|--------|--------|
| patient_1 | 0.54   | 0.32   |
| patient_2 | 0.66   | 0.10   |

In [2]:
#start code here

patients = pd.DataFrame({'gene_1':[0.54, 0.66], 'gene_2':[0.32, 0.10]}, index=['patient1', 'patient2'])

#end code here

patients

gene_1  gene_2
patient1    0.54    0.32
patient2    0.66    0.10

**question 2 (1pt)**

Use the Pandas DataFrame `.head()` function to print the first row of the `patients` DataFrame.

In [3]:
#start code here
patients.head()
#end code here

gene_1  gene_2
patient1    0.54    0.32
patient2    0.66    0.10

The following code loads the dataset we will use for this exam.

In [4]:
from sklearn.datasets import load_breast_cancer
dataset = load_breast_cancer(as_frame=True).frame
dataset.rename(columns = {'target':'label'}, inplace = True)

**question 3 (1pt)**
    
Print the number of rows and columns in the `dataset` DataFrame.

In [8]:
#start code here
rows = dataset.shape[0]
cols = dataset.shape[1]

print("Number of rows = {}".format(rows))
print("Number of columns = {}".format(cols))

#end code here

Number of rows = 569
Number of columns = 31


**question 4 (1pt)**

Print the column `label` from the `dataset` DataFrame. Don't use `.loc` or `.iloc`.

In [9]:
#start code here
print(dataset.label)
#end code here

0      0
1      0
2      0
3      0
4      0
      ..
564    0
565    0
566    0
567    0
568    1
Name: label, Length: 569, dtype: int64


**question 5 (1pt)**

Print the column `label` from the `dataset` DataFrame using `.loc`.

In [13]:
#start code here
dataset.loc[:, 'label']
#end code here

0      0
1      0
2      0
3      0
4      0
      ..
564    0
565    0
566    0
567    0
568    1
Name: label, Length: 569, dtype: int64

**question 6 (1pt)**

Print the column `label` from the `dataset` DataFrame using `.iloc`.

In [14]:
#start code here
dataset.iloc[:, -1]
#end code here

0      0
1      0
2      0
3      0
4      0
      ..
564    0
565    0
566    0
567    0
568    1
Name: label, Length: 569, dtype: int64

**question 7 (1pt)**

Print the label of the third row in the `dataset` DataFrame.

In [15]:
#start code here
dataset.iloc[2, -1]
#end code here

0

**question 8 (1pt)**

Print the subset of the `dataset` DataFrame that only contains the columns in the Python list `cols_to_print`. You have to use the `cols_to_print` list in the code that creates the subset.

In [16]:
cols_to_print = ['mean area','mean concavity','mean symmetry','label']

#start code here
dataset.loc[:, cols_to_print]
#end code here

mean area  mean concavity  mean symmetry  label
0       1001.0         0.30010         0.2419      0
1       1326.0         0.08690         0.1812      0
2       1203.0         0.19740         0.2069      0
3        386.1         0.24140         0.2597      0
4       1297.0         0.19800         0.1809      0
..         ...             ...            ...    ...
564     1479.0         0.24390         0.1726      0
565     1261.0         0.14400         0.1752      0
566      858.1         0.09251         0.1590      0
567     1265.0         0.35140         0.2397      0
568      181.0         0.00000         0.1587      1

[569 rows x 4 columns]

**question 9 (2pts)**

Print all the rows in the `dataset` DataFrame with `mean symmetry` > 0.2 and `label` equal to 1.

In [18]:
#start code here
dataset.loc[(dataset.label == 1) & (dataset.loc[:,'mean symmetry'] > 0.2)]
#end code here

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
60        10.170         14.88           64.55      311.9          0.11340   
63         9.173         13.86           59.20      260.9          0.07721   
68         9.029         17.33           58.79      250.5          0.10660   
76        13.530         10.94           87.91      559.2          0.12910   
84        12.000         15.65           76.95      443.3          0.09723   
89        14.640         15.24           95.77      651.9          0.11320   
104       10.490         19.29           67.41      336.1          0.09989   
128       15.100         16.39           99.58      674.5          0.11500   
150       13.000         20.78           83.51      519.4          0.11350   
151        8.219         20.70           53.27      203.9          0.09405   
152        9.731         15.34           63.78      300.2          0.10720   
225       14.340         13.47           92.51      641.2          0.09906   
242       11.300         18.19           73.93      389.4          0.09592   
268       12.870         16.21           82.38      512.2          0.09425   
273        9.742         15.67           61.50      289.9          0.09037   
275       11.890         17.36           76.20      435.6          0.12250   
279       13.850         15.18           88.99      587.4          0.09516   
281       11.740         14.02           74.24      427.3          0.07813   
288       11.260         19.96           73.72      394.1          0.08020   
289       11.370         18.89           72.17      396.0          0.08713   
314        8.597         18.60           54.09      221.2          0.10740   
318        9.042         18.90           60.07      244.5          0.09968   
341        9.606         16.84           61.64      280.5          0.08481   
344       11.710         15.45           75.03      420.3          0.11500   
349       11.950         14.96           77.23      426.7          0.11580   
356       13.050         18.59           85.09      512.0          0.10820   
376       10.570         20.22           70.15      338.3          0.09073   
380       11.270         12.96           73.16      386.3          0.12370   
381       11.040         14.93           70.67      372.7          0.07987   
415       11.890         21.17           76.39      433.8          0.09773   
416        9.405         21.70           59.60      271.2          0.10440   
420       11.570         19.04           74.20      409.7          0.08546   
421       14.690         13.98           98.22      656.1          0.10310   
424        9.742         19.12           61.93      289.7          0.10750   
427       10.800         21.98           68.79      359.9          0.08801   
443       10.570         18.32           66.82      340.9          0.08142   
470        9.667         18.49           61.49      289.1          0.08946   
485       12.450         16.41           82.85      476.7          0.09514   
504        9.268         12.87           61.49      248.7          0.16340   
505        9.676         13.14           64.12      272.5          0.12550   
506       12.220         20.04           79.47      453.1          0.10960   
519       12.750         16.70           82.51      493.8          0.11250   
520        9.295         13.90           59.96      257.8          0.13710   
537       11.690         24.44           76.37      406.4          0.12360   
539        7.691         25.44           48.34      170.4          0.08668   
551       11.130         22.44           71.49      378.4          0.09566   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
60            0.08061        0.010840              0.01290         0.2743   
63            0.08751        0.059880              0.02180         0.2341   
68            0.14130        0.313000              0.04375         0.2111   
76            0.10470        0

**question 10 (2pts)**

Print the median of the `area error` column in `dataset` for all rows with `label` equal to 1.

In [ ]:
#start code here

#end code here

**question 11 (3pts)**

Create a Pandes Series called `mean_area_ordinal` that, for each row in `dataset`, contains the value `low` if the value for `mean area` is within [0,420[, the value `middle` if the value for `mean area` is within [420,550[ and the value `high` if the value for `mean area` is larger of equal to 550. Use the `.apply()` function.

In [ ]:
#start code here
    
#end code here

print(mean_area_ordinal)

**question 12 (1pt)**

Write one line of code that shows how many datapoints in `mean_area_ordinal` have value `high`, how many have value `middle` and how many have value `low`.

In [ ]:
#start code here

#end code here

**question 13 (2pts)**

To prepare for Scikit-learn model fitting we need to keep the feature vectors and the label separate. Create a variable `X` that contains all the 30 features in `dataset` and a variable `y` that contains the label.

In [ ]:
#start code here

X = 
y = 

#start code here

print(X.shape)
print(y.shape)

**question 14 (1pt)**

Use the Scikit-learn `StandardScaler` to standardize the feature values in `X` (assign the result to `X_norm`).

In [ ]:
from sklearn.preprocessing import StandardScaler

#start code here

X_norm = 

#start code here

print(X_norm)

**question 15 (1pt)**

What does the following code do? What is in `X_train`, `X_test`, `y_train` and `y_test`?

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.2, random_state=0)

*Answer here*

**question 16 (2pts)**

Use the training set to compute the opimal values for the modelparameters in a Logistic Regression model (with hyperparameter `random_state` set to 1, all other hyperparameter values should be set to their Scikit-learn default values).

In [ ]:
from sklearn.linear_model import LogisticRegression

#start code here

#end code here

**question 17 (2pts)**

Given the fitted model, compute predictions on the test set and compute the Area Under the Receiver Operating Characteristic curve (AUC) for these predictions.

In [ ]:
from sklearn.metrics import roc_auc_score

#start code here

predictions = 
score_auc = 

#end code here

print(score_auc)